src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -1375,68 +1376,128 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
 }
 
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomPoint *geomPoint, PointPos PosId)
+{
+    if (PosId == PointPos::start || PosId == PointPos::mid || PosId == PointPos::end)
+        return geomPoint->getPoint();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomLineSegment *lineSeg, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return lineSeg->getStartPoint();
+    else if (PosId == PointPos::end)
+        return lineSeg->getEndPoint();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomCircle *circle, PointPos PosId)
+{
+    auto pt = circle->getCenter();
+    if(PosId != PointPos::mid)
+        pt.x += circle->getRadius();
+    return pt;
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomEllipse *ellipse, PointPos PosId)
+{
+    auto pt = ellipse->getCenter();
+    if(PosId != PointPos::mid)
+        pt += ellipse->getMajorAxisDir()*ellipse->getMajorRadius();
+    return pt;
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomArcOfCircle *aoc, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return aoc->getStartPoint(/*emulateCCW=*/true);
+    else if (PosId == PointPos::end)
+        return aoc->getEndPoint(/*emulateCCW=*/true);
+    else if (PosId == PointPos::mid)
+        return aoc->getCenter();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomArcOfEllipse *aoe, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return aoe->getStartPoint(/*emulateCCW=*/true);
+    else if (PosId == PointPos::end)
+        return aoe->getEndPoint(/*emulateCCW=*/true);
+    else if (PosId == PointPos::mid)
+        return aoe->getCenter();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomArcOfHyperbola *aoh, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return aoh->getStartPoint();
+    else if (PosId == PointPos::end)
+        return aoh->getEndPoint();
+    else if (PosId == PointPos::mid)
+        return aoh->getCenter();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomArcOfParabola *aop, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return aop->getStartPoint();
+    else if (PosId == PointPos::end)
+        return aop->getEndPoint();
+    else if (PosId == PointPos::mid)
+        return aop->getCenter();
+    return Base::Vector3d();
+}
+
+template <>
+Base::Vector3d SketchObject::getPointForGeometry<>(const Part::GeomBSplineCurve *bsp, PointPos PosId)
+{
+    if (PosId == PointPos::start)
+        return bsp->getStartPoint();
+    else if (PosId == PointPos::end)
+        return bsp->getEndPoint();
+    return Base::Vector3d();
+}

```

Comment:
All these methods seems like good candidates for switch statements.

Reply:
Done.